### ISSUE TITLE
Error dialog advises user to install a VC++ redistributable that Microsoft says is unsupported

### ISSUE THREAD

**Issue by gitname** [Vocaluxe/Vocaluxe#595]
The first time I launched `Vocaluxe.exe`, I got this error dialog:

![image](https://user-images.githubusercontent.com/7143133/214133957-1e95d4f8-ee50-4268-bfbc-c465cd796bea.png)

Here's the text (extracted from my screen using [Text Grab](https://github.com/TheJoeFin/Text-Grab)):

```text
VC++ 2010 Redistributables are missing. Please install them first.
Download(x86): https://www.microsoft.com/de-de/download/details.aspx?id=5555
Download(x64): https://www.microsoft.com/de-de/download/details.aspx?id=14632
```

I noticed a few things I consider to be UX issues:

1. The URLs are neither clickable nor selectable/copyable
2. The URLs each lead to an Error 404 page: https://www.microsoft.com/de-de/download/404Error.aspx
3. The message is in English (that's the language of my system), but the links use the `de-de` language/locale

Regarding item 2, I found this "Microsoft Visual C++ Redistributable latest supported downloads" web page:
https://learn.microsoft.com/en-US/cpp/windows/latest-supported-vc-redist

That web page lists several "C++ Redistributable" versions. It says the "2010" one is no longer supported (although it is offered for download). Given that the error message refers to the 2010 version, do you still recommend people install the 2010 version; or do you recommend they install a newer version?

---

##### Environment

- Windows 10 Enterprise LTSC 2019 (Evaluation Version)
- Vocaluxe version: `Vocaluxe_0.4.1_Windows_x64`

**Comment by Stefan1200-de** [comment]
Only Vocaluxe 0.4 needs that old Visual C++ Runtime. It is not possible to run Vocaluxe 0.4 with a newer version.

But the Vocaluxe Nightly already use newer Visual C++ runtimes, which are shipped by default with Windows 10 and newer. So feel free to upgrade to Vocaluxe Nightly, if you want to uninstall Visual C++ 2010 runtimes.

**Comment by gitname** [comment]
Thanks! Also, I have edited my original post to include the Vocaluxe version I was using (it was version 0.4, as you said).

Then, I downloaded the latest [Nightly](https://github.com/Vocaluxe/Vocaluxe/releases/tag/Nightly) build (i.e. commit https://github.com/Vocaluxe/Vocaluxe/commit/dbc1738da1c848facb7be6b5596ebad8b0ba9f44 dated Oct 29, 2022) and got a similar error when I launched `Vocaluxe.exe`. Here's a screenshot:

![image](https://user-images.githubusercontent.com/7143133/214188855-151738df-97e1-428c-8056-5841e00f1d4b.png)

The text in the upper field is:

```text
VC++ 2010 Redistributables are missing. Please install them first. 
Download(x86): https://www.microsoft.com/de-de/download/details.aspx?id=5555 
Download(x64): https://www.microsoft.com/de-de/download/details.aspx?id=14632
```

And the text in the lower field is:

```text
[00:00:00.0239984] [Information] Starting to log

{ThreadId: 1, AdditionalData: ["{ Version = Vocaluxe 'Blue Sunrise' 0.4.0 (x64) (0.4.0-605-gdbc1738) }", "{ StartDate = 1/23/2023 4:45:26 PM }", "{ Id = 56752bb3-ca55-47f7-9df0-f6d2195a5a18 }"], CallingContext: "{ callerMethodeName = Init, callerFilePath = C:\projects\vocaluxe\VocaluxeLib\Log\CLog.cs, callerLineNumer = 128 }"}

[00:00:00.0479998] [Fatal] VC++ 2010 Redistributables are missing. Please install them first. 

Download(x86): https://www.microsoft.com/de-de/download/details.aspx?id=5555 

Download(x64): https://www.microsoft.com/de-de/download/details.aspx?id=14632

{ThreadId: 1, CallingContext: "{ callerMethodeName = CheckRequirements, callerFilePath = C:\projects\vocaluxe\Vocaluxe\CProgrammHelper.cs, callerLineNumer = 122 }"}
```

Looks like a similar error to me as with Vocaluxe 0.4. I wonder whether _LTSC_ versions of Windows Enterprise don't include the necessary C++ redistributable. (I could download and install one from the Microsoft web page I linked to earlier.)

---

##### Environment
- Windows 10 Enterprise LTSC 2019 (Evaluation Version)
    ```shell
    $ systeminfo
    # ...
    OS Name:                   Microsoft Windows 10 Enterprise LTSC Evaluation
    OS Version:                10.0.17763 N/A Build 17763
    ``` 
- Vocaluxe version: latest Nightly build (i.e. commit https://github.com/Vocaluxe/Vocaluxe/commit/dbc1738da1c848facb7be6b5596ebad8b0ba9f44 dated Oct 29, 2022)

**Comment by gitname** [comment]
I just saw this in `Readme.txt` (in the downloaded ZIP file):

```yaml
Supported Operating Systems / Requirements:
- Windows Vista, Windows 7, Windows 8 or Windows 10 with .NET 4.0
- Linux (no official support)
- 1 GHz CPU, 512 MB RAM, Graphics card with OpenGL 2.1 or DirectX 9.0 support
- Visual C++ Redistributable Package 2010 (2015-2022 is required for the nightly builds)
```

I will try downloading and installing that specific Visual C++ Redistributable and report back.

**Comment by gitname** [comment]
I downloaded and installed the **Microsoft Visual C++ 2015-2022 Redistributable (x64) - 14.34.31931** from here: https://aka.ms/vs/17/release/vc_redist.x64.exe (which is a link I got from here: https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170#visual-studio-2015-2017-2019-and-2022). Then, I restarted my PC as instructed by the installer.

With that installed (and after restarting my PC), I launched `Vocaluxe.exe` (Nightly build) once again. I got the same error dialog:

![image](https://user-images.githubusercontent.com/7143133/214193179-e6fea72a-59bc-482e-bb9c-b17e67970c27.png)

The error message still refers to a "2010" version of the redistributable.

**Comment by Stefan1200-de** [comment]
That surprise me a little bit. But I need the Visual C++ runtime also for another applications, so I never did tests without it.

**Comment by gitname** [comment]
Oh, OK. 

I searched the repository for the string "redistributable" and found some code that checks whether a program named "Microsoft Visual C++ 2010" is installed.

- The [_Run](https://github.com/Vocaluxe/Vocaluxe/blob/dbc1738da1c848facb7be6b5596ebad8b0ba9f44/Vocaluxe/Program.cs#L84) function (unconditionally) invokes...
- The [CheckRequirements](https://github.com/Vocaluxe/Vocaluxe/blob/dbc1738da1c848facb7be6b5596ebad8b0ba9f44/Vocaluxe/CProgrammHelper.cs#L117) function, which (within an `#if WIN ... #endif` block) invokes...
- The [_IsVC2010Installed](https://github.com/Vocaluxe/Vocaluxe/blob/dbc1738da1c848facb7be6b5596ebad8b0ba9f44/Vocaluxe/CProgrammHelper.cs#L94) function, which (unconditionally) invokes...
- `_IsProgramInstalled("Microsoft Visual C++ 2010")`

I don't have much experience reading/writing C#. I assume I could remove the invocation of the `_IsVC2010Installed` function, re-build the program, and run it, and that would work around the error I am seeing (although I don't want to do that if Vocaluxe really needs the "2010" version for something). I don't want to install the entire Visual Studio onto my system, but I do have Visual Studio Code installed and will look into the possibility of building Vocaluxe [using that](https://code.visualstudio.com/docs/languages/csharp) instead.

**Comment by gitname** [comment]
> I do have Visual Studio Code installed and will look into the possibility of building Vocaluxe [using that](https://code.visualstudio.com/docs/languages/csharp) instead.

I installed the [C# Language Extension for VS Code](https://code.visualstudio.com/docs/languages/csharp). Then, in VS Code, I performed an "OmniSharp: Restart OmniSharp" operation. I got an error message (I didn't record it) and the OmniSharp log said that my system lacked [something related to building things for] .NET version 4.7.1.

I then downloaded the [.NET Framework 4.7.1 Developer Pack Offline Installer](https://dotnet.microsoft.com/en-us/download/dotnet-framework/net471) and installed that. Then, in VS Code, I performed another "OmniSharp: Restart OmniSharp" operation. I got the error message "Some projects have trouble loading. Please review the output for more details." (This may have been the same error message as I mentioned in the previous paragraph). This time, the OmniSharp log did not say my system lacked the .NET 4.7.1-related things. But, it did show the following errors:

```js
[fail]: OmniSharp.MSBuild.ProjectLoader
        This project references NuGet package(s) that are missing on this computer. Rebuild for automatic package restore. The missing file is ..\packages\vocaluxe.dependencies.acinerella.1.0.6\build\vocaluxe.dependencies.acinerella.targets.

// ...

[warn]: OmniSharp.MSBuild.ProjectManager
        Failed to load project file 'c:\Temp\Code\_contributions\Vocaluxe\Vocaluxe\Vocaluxe.csproj'.
c:\Temp\Code\_contributions\Vocaluxe\Vocaluxe\Vocaluxe.csproj
c:\Temp\Code\_contributions\Vocaluxe\Vocaluxe\Vocaluxe.csproj(567,5): Error: This project references NuGet package(s) that are missing on this computer. Rebuild for automatic package restore. The missing file is ..\packages\vocaluxe.dependencies.acinerella.1.0.6\build\vocaluxe.dependencies.acinerella.targets.

// ...

[fail]: OmniSharp.MSBuild.ProjectLoader
        This project references NuGet package(s) that are missing on this computer. Rebuild for automatic package restore. The missing file is ..\packages\vocaluxe.dependencies.securitycryptography.1.0.1\build\vocaluxe.dependencies.securitycryptography.targets.
[warn]: OmniSharp.MSBuild.ProjectManager
        Failed to load project file 'c:\Temp\Code\_contributions\Vocaluxe\WebserverInitalConfig\WebserverInitalConfig.csproj'.
c:\Temp\Code\_contributions\Vocaluxe\WebserverInitalConfig\WebserverInitalConfig.csproj
c:\Temp\Code\_contributions\Vocaluxe\WebserverInitalConfig\WebserverInitalConfig.csproj(76,5): Error: This project references NuGet package(s) that are missing on this computer. Rebuild for automatic package restore. The missing file is ..\packages\vocaluxe.dependencies.securitycryptography.1.0.1\build\vocaluxe.dependencies.securitycryptography.targets.

// ...

[fail]: OmniSharp.MSBuild.ProjectLoader
        This project references NuGet package(s) that are missing on this computer. Use NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is ..\packages\NUnit.3.10.1\build\NUnit.props.
[warn]: OmniSharp.MSBuild.ProjectManager
        Failed to load project file 'c:\Temp\Code\_contributions\Vocaluxe\Tests\Tests.csproj'.
c:\Temp\Code\_contributions\Vocaluxe\Tests\Tests.csproj
c:\Temp\Code\_contributions\Vocaluxe\Tests\Tests.csproj(120,5): Error: This project references NuGet package(s) that are missing on this computer. Use NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is ..\packages\NUnit.3.10.1\build\NUnit.props.

// ...
```

This is involving installing more software onto this system than I am comfortable. I'll look into setting up a different computer where I am comfortable installing the "Microsoft Visual C++ 2010 Redistributable" (the one Microsoft says it does not support anymore). I just don't want it on this computer, which is my main one. I have an old laptop I can experiment with.

**Comment by lukeIam** [comment]
acinerella/ffmpeg needs a vc++ runtime (so vocaluxe does also).
But it can be compiled also with newer versions (i updated it long time
ago).
But don't ask me what the current build is (but i can figure this out)

**Comment by marwin89** [comment]
@gitname, do you still got this problem? 👋🏻 

Maybe you just uninstall everything and follow [Installation Guide](https://github.com/Vocaluxe/Vocaluxe/wiki/Installation) (in Wiki). I updated it last week. There is a [new nightly build from 10 January 2024](https://www.vocaluxe.org/#download) you might wanna try.

BTW: we're always [looking for web developers](https://github.com/orgs/open-music-games/discussions/24) for all open-music-games projects (Vocaluxe, Performous and more). 🙂

**Comment by marwin89** [comment]
I close this issue due inactivity. Please re-open if needed.

**Comment by gitname** [comment]
Sorry about the delay in my response, everyone. Thanks for getting back to me.

I have not tried using this program since I created this Issue back in January, 2023. As a result, I don't have any news to report. Due to a change in my priorities since then, I don't have any immediate plans to use this program. I just wanted to follow up here and express my appreciation for you following up with me above. No need to reply.

On a related note, I no longer use Windows 10 (I have upgraded to Windows 11) and so will not be able to practically re-test the exact same configuration as I initially reported the issue under.
